### PR TITLE
HIVE-25441: Incorrect deltas split for sub-compactions when using

### PR DESCRIPTION
**Pull request by deniskuzZ**

### CODE REVIEW

**Comment by deniskuzZ** [comment]
recheck

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java [hunk 1]
```diff
@@ -234,12 +234,24 @@ void run(HiveConf conf, String jobName, Table t, Partition p, StorageDescriptor
...
         "runaway/mis-configured process writing to ACID tables, especially using Streaming Ingest API.");
       int numMinorCompactions = parsedDeltas.size() / maxDeltasToHandle;
+      parsedDeltas.sort(AcidUtils.ParsedDeltaLight::compareTo);
+
+      int start = 0;
+      int end = maxDeltasToHandle;
+
       for (int jobSubId = 0; jobSubId < numMinorCompactions; jobSubId++) {
+        while (parsedDeltas.get(end).getMinWriteId() == parsedDeltas.get(end - 1).getMinWriteId() &&
+          parsedDeltas.get(end).getMaxWriteId() == parsedDeltas.get(end - 1).getMaxWriteId()) {
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:244]: Sorry I don't get this part. What does it do?

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:244]: it removes deltas for statements that went of the batch size. ( [... delta_5_5_1 ] [delta_5_5_2, ...])  - we should process multiple statements for the same writeIds range in the same batch, in this scenario delta_5_5_1 should be included in the next batch.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 2]
```diff
@@ -781,6 +782,67 @@ public void autoCompactOnStreamingIngestWithDynamicPartition() throws Exception
...
+
+    for (FileStatus f : stat) {
+      for (int j = 0; j < matchesNotFound.size(); j++) {
+        if (f.getPath().getName().matches(matchesNotFound.get(j))) {
+          matchesNotFound.remove(j);
+          break;
+        }
+      }
+    }
+    Assert.assertEquals("Matches Not Found: " + matchesNotFound.toArray(), 0, matchesNotFound.size());
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java]: fixed

**Review by klcopp — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java (hunk 1 above)
Comment (anishek) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:244]: with hive.compactor.delta.pct.threshold=2 and above code would it give an error ?

Comment (anishek) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:244]: actually thinking some more:: if there are multiple deltas with same min,max and different statement id then we have to make sure that value of hive.compactor.delta.pct.threshold is greater than any length of such sorted list or we have to artificially at that point increase the above value with a warning in log

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:244]: 'hive.compactor.delta.pct.threshold' is just one of the conditions for auto-compaction.

**Pull request merged**